Photon updates, 2nd pass

devinreams opened the pull request:
Resolves #331 

~Starting a branch off of the WIP doorhanger #338 to work on style improvements~

- [x] Navigation links (top right) to match the photon guidelines, found [here](http://design.firefox.com/photon/components/link.html) including external link icon
- [x] Vertical height of Nav bar, 72px
- [x] Moving entries counter into primary content area (You have __ entries in your Lockbox)
- [x] Moving home next to "New Entry" and applying secondary button styling to it
- [x] Sidebar width (320px)
- [x] Styling of new entry in sidebar to match with content background of main content area...
- [x] Removal of horizontal divider in Create a New Entry & View Entry Details
- [x] Change background color of primary content area when viewing or creating an entry to hex F3F3F5, as seen in latest design comps. 
- [x] Fix password and search input font color so its legible (very light, reported #264)
- [x] 1.1.0 welcome screen width and fonts
- [x] revisit button sizes
- [x] "Site Address" and "Site Name" order and copy
- [x] sidebar: border on all items except when item selected (and no arrow)
- [x] sidebar and doorhanger nav: Photon icon ["Arrowhead Right 16"](http://design.firefox.com/icons/viewer/)
- [x] sidebar: zebra background colors conform to comps
- [x] modal copy changes for delete
- [x] modal rounded with subtle-er shadow
- [x] entry details: background color 
- [x] borrow content from welcome or otherwise and drop into addon description

----

## Review Comments

- [x] `list/manage/components/app.css` "While 320px is fine as a target, I'd expect this to flex smaller for narrower windows." recommend something more modest, then adjust the grid-template-columns property of the parent node
- [x] `list/manage/components/item-details.css` Dark list items should be Grey 20 (#EDEDF0), Lighter list items should be Grey 10 (#F9F9FA)
  - ~`<ItemDetails/>` (and `<EditItemDetails/>`) are designed as (theoretically) reusable components that could be put anywhere. Somewhere in `<CurrentSelection/>` would be a better place~ moved to #392
  - ~(NEW) Content area bg for the selected item should match either of these two values depending on the color of the list item selected~ moved to #392
- [x] `list/manage/components/item-details.css` height: 100% "I don't think this does what you want"
- [x] `list/manage/components/item-details.css` border-bottom "What is this for?"
- [x] `list/manage/containers/item-count.js ` toolbar-item-count "Don't forget to remove the corresponding l10n string."
- [ ] `package.json` description "this doesn't follow the Photon spec on capitalization..." 
  - [x] and make consistent in all places..
- [x] `list/components/item-summary.css` arrowhead-right "This isn't vertically centered"
- [x] `list/manage/components/homepage.js` h1 "You don't need the {""} surrounding this text."
- [x] `locales/en-US/list.ftl` "Provide Feedback"
- [x] `widgets/button.css` normal-theme:hover "This rule isn't redundant.."
- [x] `widgets/button.css` min-width is property of size, not theme...
- [x] `widgets/button.js` "Rather than making a new theme, I think we want to create a whole new `<Link/>` component" handled by #397
  - ~"in the short term, we'd want to apply the correct ARIA role to link-themed buttons"~ superseded
- [x] `widgets/dialog-box.css` font-width: 300 "This seems a bit big/heavy for the dialog text... not so bad for delete"
  - [x] "relative font sizes emphasize the description of the dialog and not the buttons"
- [x] `widgets/panel.js` "We should localize the alt to say something like "Go back"."
- [x] fix integration tests

devinreams commented:
For those following along, here was the result of a very quick-n-dirty pass at the punch-list from #331:

![screen shot 2017-11-22 at 9 51 41 am](https://user-images.githubusercontent.com/49511/33139640-23db5f96-cf6b-11e7-9eae-63f20c1cb947.png)

I know @changecourse is still finalizing some of the screens and planning to get them into Zeplin so we can confirm sizes, shapes, colors, etc. fairly quickly and easily. So, I'll hit pause 🙅‍♂️ here for a little bit knowing we're moving in a good direction and no big concerns spotted (yet?).

For what its worth, the refactor and work @jimporter has done in #338 made the shared components (and styles) very easy to target and update. 👏

pdehaan commented on src/webextension/list/components/item-summary.css:
```diff
@@ -5,15 +5,27 @@
 }
 
 .title, .subtitle {
+  weight: 600;
```
typo: `font-weight`

devinreams replied:
Thanks very much @pdehaan 🙇

pdehaan commented on src/webextension/list/components/item-summary.css:
```diff
@@ -5,15 +5,27 @@
-  font-size: 80%;
-  opacity: 0.75;
+  font-size: 13px;
+  font-wight: 400;
```
typo: `font-weight`

devinreams commented:
⚠️ Note: I re-did this branch and force pushed now that the doorhanger has merged... just in case anything looked weird to anyone coming through here.

devinreams commented:
This PR is getting a bit .. long.

@jimporter Can you please help me review it as-is and then work on a follow-up branch to tackle some of the more interesting/tricky stuff outlined below?

@linuxwolf I will still assume I should hold off on a merge here until your branch has landed since so much moved around on the popup, homepage, first run, etc.

## DONE (ready for review)

- [x] Navigation links (top right) to match the photon guidelines, found [here](http://design.firefox.com/photon/components/link.html) including external link icon
- [x] Vertical height of Nav bar, 72px
- [x] Moving entries counter into primary content area (You have __ entries in your Lockbox)
- [x] Moving home next to "New Entry" and applying secondary button styling to it
- [x] Sidebar width (320px)
- [x] Styling of new entry in sidebar to match with content background of main content area, based on screenshot below: <img width="583" alt="screen shot 2017-11-13 at 11 43 35 am" src="https://user-images.githubusercontent.com/6164801/32742898-ff9c4cee-c867-11e7-9e80-7bbce505f46b.png">
- [x] Removal of horizontal divider in Create a New Entry & View Entry Details<img width="385" alt="screen shot 2017-11-13 at 11 44 51 am" src="https://user-images.githubusercontent.com/6164801/32742940-20ba234c-c868-11e7-9223-3c0e8793316a.png">
- [x] Change background color of primary content area when viewing or creating an entry to hex F3F3F5, as seen in latest design comps. 
- [x] Fix password and search input font color so its legible (very light, reported #264)
  <img width="681" alt="screen shot 2017-11-14 at 10 09 03 am" src="https://user-images.githubusercontent.com/49511/32793670-de8343b6-c923-11e7-9f55-b02f404e6f03.png">
- [x] 1.1.0 welcome screen width and fonts
- [x] revisit button sizes
- [x] "Site Address" and "Site Name" order and copy
- [x] sidebar: border on all items except when item selected (and no arrow)
- [x] sidebar and doorhanger nav: Photon icon ["Arrowhead Right 16"](http://design.firefox.com/icons/viewer/)
- [x] sidebar: zebra background colors conform to comps
- [x] modal copy changes for delete
- [x] modal rounded with subtle-er shadow
- [x] entry details: background color 
- [x] borrow content from welcome or otherwise and drop into addon description

## TODO (candidates for follow-up)

- [ ] extend background of item details all the way to bottom of screen (content area is gray, white space beneath)
- [ ] zero entries / first run home page with 3-up feature heros
- [ ] sidebar: secondary (no username)" and "(no site name)" CSS treatment (lighter color)
- [ ] "Disabled"/inactivate/muted state of the sidebar list items when an item is selected (greys out all entries but the one selected)
- [ ] Search bar magnify icon ~and clear icon (when populated)~ based on screenshot below: <img width="260" alt="screen shot 2017-11-13 at 11 41 21 am" src="https://user-images.githubusercontent.com/6164801/32742816-b9918336-c867-11e7-995a-8de86e47d694.png">
- [ ] red alert button style for delete (keep primary for discard)
- [ ] a little wider width for doorhanger?
- [ ] "manage lockbox" button shorter, background color? (instead of ghost button, make it a puffy?)
- [ ] entry details: move "title" to title bar (out of detail view) and shorter
- [ ] doorhanger search padding to have more space
- [ ] review all copy text once more..

linuxwolf commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "version": "0.1.3-alpha",
   "main": "dist/bootstrap.js",
-  "description": "A Lockbox extension for Firefox",
+  "description": "the simple way to store, retrieve and manage website login info",
```
RIP, Oxford Comma

jimporter replied:
Mentioned in a previous comment, but this doesn't follow the Photon spec on capitalization.

devinreams commented:
This has been rebased (and resolved) against all the work at #362 (FxA) and force pushed. ✅ 

@jimporter This is ready for your review today knowing there are a bunch of "todo" items still. I want to try to tackle those in (a) separate small branch(es) to try and get as much done as quickly before we have to start our QA and release process.

devinreams commented:
Integration tests are failing but re-written at #391. Once both have landed this will look a bit more sane.

devinreams commented:
I'm going to 🔨 call this branch good-enough and spin up a "3rd pass" issue to track the remaining open items so we can land this as-is: https://github.com/mozilla-lockbox/lockbox-extension/issues/392

devinreams commented:
@jrbenny35 I somehow broke these integration when merging in the latest (they were passing before). I've looked around at the sign in selectors and I think I updated them to be accurate (last 2 commits) but I'm still getting failures on the last three tests:

https://travis-ci.org/mozilla-lockbox/lockbox-extension/jobs/313141734#L2027-L2029

You're much smarter/faster at these than I. Can you tell what the root cause is here? 😬

jrbenny35 commented:
@devinreams looks like you changed it to ```puffy-size``` where-as in the ```util.py``` file it is ```puffy-theme```.

jimporter commented on src/webextension/list/manage/containers/item-count.js:
```diff
@@ -1,28 +0,0 @@
-
-function ItemCount({count}) {
-  return (
-    <Localized id="toolbar-item-count" $count={count}>
```
Don't forget to remove the corresponding l10n string.

jimporter commented on src/webextension/locales/en-US/list.ftl:
```diff
@@ -52,11 +52,7 @@ all-items-empty =
-     [1]     You have { $count } entry in your { product-title }
-    *[other] You have { $count } entries in your { product-title }
-  }
+  the simple way to store, retrieve, and manage website login info
```
I think I commented on this on #362, but this doesn't appear to follow the Photon [spec for capitalization](https://design.firefox.com/photon/copy/capitalization.html). I looked around through the Firefox UI and I can't find any examples of all lower-case text like this. Pretty much everything is either title case or sentence case (with a smattering of all-caps stuff).

jimporter commented on src/webextension/locales/en-US/list.ftl:
```diff
@@ -54,19 +54,11 @@ all-items-filtered = No results
 homepage-linkaccount-description =
-    Creating a Firefox account - or adding { product-title } to an existing account - protects your logins
-    with the strongest encryption available.
+    Now create a Firefox account – or add Lockbox to an existing
```
Should there be spaces around the em dash? Usually I see them used without spaces.

devinreams commented on src/webextension/list/manage/components/item-details.css:
```diff
@@ -3,7 +3,10 @@
 .item-details {
-  margin: 2em;
+  padding: 2em;
+  background-color: #f3f3f5;
```
>It's not one of the defined colors in the Photon theme specs

@changecourse can you comment on this? 😕 

>I don't think this is the right place to put these styles; <ItemDetails/> (and <EditItemDetails/>) are designed as (theoretically) reusable components that could be put anywhere.

👍

changecourse replied:
Agreed... updated colors to align to spec. 

Dark list items should be Grey 20 (#EDEDF0)
Lighter list items should be Grey 10 (#F9F9FA)

Content area bg for the selected item should match either of these two values depending on the color of the list item selected.

devinreams commented on src/webextension/list/manage/components/item-details.css:
```diff
@@ -3,7 +3,10 @@
-  margin: 2em;
+  padding: 2em;
+  background-color: #f3f3f5;
+  height: 100%;
```
>The spec shows that all the empty space below the form should be the same color, but in practice, the background-color defined above only applies to the form area.

You're right. And I struggled to see why/how to fix it with a quick glance. I assume it has to do with grid/flex? If you have any pointers I'd appreciate it, otherwise I'll keep hunting around.

devinreams replied:
>And I struggled to see why/how to fix it with a quick glance. I assume it has to do with grid/flex?

Wait, never mind, I think I see it now. Removing the "height" and set the right background-color.

devinreams commented on src/webextension/list/components/item-fields.js:
```diff
@@ -46,20 +46,20 @@ const fieldsPropTypes = PropTypes.shape({
       <div className={styles.field}>
         <Localized id="item-fields-origin">
-          <LabelText>oRIGIn</LabelText>
+          <LabelText className={styles.firstLabel}>oRIGIn</LabelText>
```
>What's the reasoning behind this?

@changecourse can better explain the re-ordering of Site Address / Site Name in context of Jim's questions and comments. ☝️

changecourse replied:
The reordering was based on user understanding of the entry fields, as part of the alpha evaluative testing sessions... Participants were generally confused about "Entry/Site name" as the first item... and by having web address first, we provide some good context (by order of operations) for someone to then understand what the following "Site Name" is.

jimporter replied:
That doesn't address all of my concerns, unfortunately. As mentioned, we plan to support multiple origins (we'll need them as soon as we start doing, well, *anything* with the origin :)), and I'm not sure an editable list of origins makes sense as the first field. I suppose we could change it back when we get that, but given that the current origin field is very much temporary, I'm not sure it's worth worrying about now...

devinreams commented on src/webextension/list/manage/components/send-feedback.js:
```diff
@@ -18,7 +18,7 @@ export default function SendFeedback() {
   return (
     <Localized id="toolbar-send-feedback">
-      <Button theme="ghost" onClick={doClick}>
+      <Button theme="ghost" size="puffy" onClick={doClick}>
```
>Have we run this by the Photon design folks? My reading of the Photon spec doesn't indicate to me that this is the right place to use puffy buttons.

@changecourse As I'm following your toolbar button designs (more or less successfully...) can you comment on the above? "Don't worry about it.." may be an acceptable response. 😏

changecourse replied:
I've gone back n' forth on this, but landed on puffy to be painfully obvious for primary/secondary CTA's for the time being... we may revisit and make them smaller down the road if needed.

devinreams commented on src/webextension/locales/en-US/list.ftl:
```diff
@@ -46,10 +46,10 @@ toolbar-item-count =
 toolbar-add-item = New Entry
 toolbar-go-home = Home
-toolbar-send-feedback = Provide Feedback
+toolbar-send-feedback = Submit Feedback
```
>"Submit Feedback" really sounds like the thing that happens after you've filled out the feedback form and are sending it to us, especially since "Submit" is a somewhat-technical term

@changecourse I presume "Submit" was an intentional decision made recently?

changecourse replied:
This was a copy request from our team, but I tend to agree with you... I'm not happy about "Submit" as opposed to "Provide"

@sandysage final call.

sandysage replied:
Let's stick with "provide" as @changecourse suggests.

devinreams commented on src/webextension/widgets/button.css:
```diff
@@ -5,15 +5,19 @@
   white-space: nowrap;
   text-align: center;
+  letter-spacing: 0.2px;
+  line-height: 1.5;
```
>Where are you getting these values? They don't appear to be in the Photon spec.

@changecourse described today in our call some of the thinking of a first pass around typography spacing. Ryan, I assume we keep this as-is? Any more comments or context to share?

changecourse replied:
Agreed. 1.5x the font size for line-height is a readability standard, not currently covered in the spec, but a candidate for future inclusion. 

0.2px letter-spacing is something I could go either way on, but I have found legibility success with this on many past projects and would like to leave it in for the time being.

devinreams commented:
Now that master is... very different. I'm going to:

1. pull out a punch-list of review todos and add them to the original description above
2. cherry-pick _just_ my work here onto a branch and force-push it over this one
3. work on the review todos and check them off above, with one commit per todo

Because of 2, review stuff will kinda disappear (thanks, GitHub), hence step 1. But then it'll be nice clean history (without merge and revert commits), as @jimporter requested and hopefully easy to re-review.

Stay tuned...

devinreams commented:
Here's my todo list based on the decisions made thus far above:

- [ ] `list/manage/components/app.css` "While 320px is fine as a target, I'd expect this to flex smaller for narrower windows." recommend something more modest, then adjust the grid-template-columns property of the parent node
- [ ] `list/manage/components/item-details.css` Dark list items should be Grey 20 (#EDEDF0), Lighter list items should be Grey 10 (#F9F9FA), Content area bg for the selected item should match either of these two values depending on the color of the list item selected
- [x] `list/manage/components/item-details.css` height: 100% "I don't think this does what you want"
- [x] `list/manage/components/item-details.css` border-bottom "What is this for?"
- [x] `list/manage/containers/item-count.js ` toolbar-item-count "Don't forget to remove the corresponding l10n string."
- [ ] `package.json` description "this doesn't follow the Photon spec on capitalization..." (and make consistent in all places..)
- [ ] `list/components/item-summary.css` arrowhead-right "This isn't vertically centered"
- [x] `list/manage/components/homepage.js` h1 "You don't need the {""} surrounding this text."
- [ ] `locales/en-US/list.ftl` "Provide Feedback"
- [ ] `widgets/button.css` normal-theme:hover "This rule isn't redundant.."
- [ ] `widgets/button.css` min-width is property of size, not theme...
- [ ] `widgets/button.js` "Rather than making a new theme, I think we want to create a whole new <Link/> component"
  - [ ] "in the short term, we'd want to apply the correct ARIA role to link-themed buttons"
- [ ] `widgets/dialog-box.css` font-width: 300 "This seems a bit big/heavy for the dialog text... not so bad for delete"
  - [ ] "relative font sizes emphasize the description of the dialog and not the buttons"
- [ ] `widgets/panel.js` "We should localize the alt to say something like "Go back"."

Here we goooooooo.... (in Mario voice)

devinreams commented:
OK, cherry-pick is complete. I didn't bring over the `ToolbarSpace` commit/revert noise. Nor my flurry of attempts to fix integration tests. ❌ 

Onward to review comments...

devinreams commented:
Alright I've tackled more changes and pushed the latest. ☝️ 

I saved the best for last (todo tomorrow first thing):

- [ ] `list/components/item-summary.css` arrowhead-right "This isn't vertically centered"
- [ ] `widgets/button.js` "Rather than making a new theme, I think we want to create a whole new `<Link/>` component"
- [ ] fix integration tests ⚠️ 


@jrbenny35 I think I've messed up my local tox because now I'm getting these errors (even on master after npm install) below. Did I need to do something to update tox, dependencies, or something else? If you're willing/able to help troubleshoot I'll be on Slack and I'll be very grateful for your help again. :)

<details>

```
py36-integration-tests installed: apipkg==1.4,certifi==2017.11.5,chardet==3.0.4,execnet==1.5.0,idna==2.6,py==1.5.2,PyPOM==1.2.0,pytest==3.2.2,pytest-base-url==1.4.1,pytest-html==1.16.0,pytest-metadata==1.5.1,pytest-selenium==1.11.0,pytest-variables==1.7.0,pytest-xdist==1.18.2,requests==2.18.4,selenium==3.6.0,urllib3==1.22,zope.component==4.4.1,zope.event==4.3.0,zope.interface==4.4.3
py36-integration-tests runtests: PYTHONHASHSEED='3398831235'
py36-integration-tests runtests: commands[0] | pytest --driver=Firefox --verbose test/integration
Traceback (most recent call last):
  File "/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/lib/python3.6/site-packages/_pytest/config.py", line 342, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/dreams/src/lockbox-extension/test/integration')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/lib/python3.6/site-packages/_pytest/config.py", line 373, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/Users/dreams/src/lockbox-extension/test/integration/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/lib/python3.6/site-packages/_pytest/config.py", line 379, in _importconftest
    mod = conftestpath.pyimport()
  File "/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/lib/python3.6/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/lib/python3.6/site-packages/_pytest/assertion/rewrite.py", line 212, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/Users/dreams/src/lockbox-extension/test/integration/conftest.py", line 6, in <module>
    from fxapom.fxapom import PROD_URL, FxATestAccount
ModuleNotFoundError: No module named 'fxapom'
ERROR: could not load /Users/dreams/src/lockbox-extension/test/integration/conftest.py

ERROR: InvocationError: '/Users/dreams/src/lockbox-extension/.tox/py36-integration-tests/bin/pytest --driver=Firefox --verbose test/integration'
flake8 installed: apipkg==1.4,certifi==2017.11.5,chardet==3.0.4,configparser==3.5.0,enum34==1.1.6,execnet==1.5.0,flake8==3.5.0,flake8-docstrings==1.1.0,flake8-isort==2.2.2,flake8-polyfill==1.0.1,idna==2.6,isort==4.2.15,mccabe==0.6.1,py==1.5.2,pycodestyle==2.3.1,pydocstyle==2.1.1,pyflakes==1.6.0,PyPOM==1.2.0,pytest==3.2.2,pytest-base-url==1.4.1,pytest-html==1.16.0,pytest-metadata==1.5.1,pytest-selenium==1.11.0,pytest-variables==1.7.0,pytest-xdist==1.18.2,requests==2.18.4,selenium==3.6.0,six==1.11.0,snowballstemmer==1.2.1,testfixtures==5.3.1,urllib3==1.22,zope.component==4.4.1,zope.event==4.3.0,zope.interface==4.4.3
flake8 runtests: PYTHONHASHSEED='3398831235'
flake8 runtests: commands[0] | flake8 .
./test/integration/conftest.py:6:1: I003 isort expected 1 blank line in imports, found 0
./test/integration/conftest.py:7:1: I004 isort found an unexpected blank line in imports
ERROR: InvocationError: '/Users/dreams/src/lockbox-extension/.tox/flake8/bin/flake8 .'
_______________________________________________________ summary ________________________________________________________
ERROR:   py36-integration-tests: commands failed
ERROR:   flake8: commands failed
```

</details>

jrbenny35 commented:
@devinreams try this: ```tox -r```. ```-r``` is recreate and will download all the current requirements again. Since ```fxapom``` was just added, it needs to be download into your local tox directory.

devinreams commented:
>-r is recreate and will download all the current requirements again

Aha! That's the ticket, thank you! I was `pip install`ing fxapom and trying everything else I could think of on the python side. 😫

devinreams commented:
@linuxwolf I'm struggling a bit on how to fix the `fxa_sign_in` tests since the design has changed a bit (see 4542ef1):

- The button on the firstrun page is now a "link-theme" with "external" class
  ![screen shot 2017-12-08 at 10 58 15 am](https://user-images.githubusercontent.com/49511/33778643-ca8684e0-dc06-11e7-9853-cf5811e7cd6f.png)

- Whereas the homepage is still intended to be a big puffy button

That all said, @jimporter's feedback was to create a `<Link/>` component to replace my "link-theme". Is this something you'd be able and willing to help tackle in this branch today? I'm a bit stuck 😬

linuxwolf commented:
> That all said, @jimporter's feedback was to create a <Link/> component to replace my "link-theme". Is this something you'd be able and willing to help tackle in this branch today? I'm a bit stuck 😬

I'll see what I can do.  It doesn't seem _too_ bad ...

linuxwolf commented on src/webextension/widgets/button.js:
```diff
@@ -11,6 +11,7 @@ const THEME_CLASS_NAME = {
   primary: `${styles.primaryTheme}`,
   normal: `${styles.normalTheme}`,
   ghost: `${styles.ghostTheme}`,
+  link: `${styles.linkTheme}`,
```
in the PR, the "link" theme is being used for:

* The welcome page's "Sign in to your Firefox Account" button (lots of logic back there)
* The "Copy \[to clipboard]" button
* The "Send Feedback" (open an external page)

Of these, one is definitely open an external link.  The others are really actions.  Is it appropriate to treat all of these treated as external links, even for ARIA?

jimporter replied:
The only one that we want to use `<Link/>` now is "Send Feedback"; the others are really buttons. While the theming for them might want to change eventually (especially for the Copy button), I don't think we should make things that are actions look like links, for the same reason we wouldn't give them an ARIA role of "link".

devinreams replied:
>The only one that we want to use <Link/> now is "Send Feedback"

Agreed. And the second will be the "FAQ" link (update needed after this is completed) over at #345

devinreams commented:
@jimporter I re-read your comment around the background-color: "`<ItemDetails/>` (and `<EditItemDetails/>`) are designed as (theoretically) reusable components that could be put anywhere. Somewhere in `<CurrentSelection/>`..". Is this still worth addressing now or perhaps later given @changecourse's new feature request "Content area bg for the selected item should match either of these two values depending on the color of the list item selected".

I propose we move these two bits of feedback/functionality over to #392 to tackle later and leave `item-details.css` as-is. Is that OK?

jimporter commented:
Yeah, definitely do that later. I'm not even sure that's something we should do at all, since it affects the contrast of the text, which can pose an accessibility problem. WCAG has some pretty stringent guidelines about the minimum contrast level of text.

devinreams commented:
>Yeah, definitely do that later. I'm not even sure that's something we should do at all, since it affects the contrast of the text

@jimporter OK cool. I moved both for further discussion later.

Would you mind taking another look at this now? (we can assume #397 will address your comments and merge in shortly)

My work after your review starts at https://github.com/mozilla-lockbox/lockbox-extension/pull/351/commits/a7133d69358f24355efcae1b0e276e1d8dbade17

linuxwolf commented:
@devinreams @jrbenny35 my last commit fixes _one_ of the problems after the refactor.  Still getting errors because selectors don't line up.

jimporter commented on src/webextension/widgets/input.css:
```diff
@@ -78,17 +79,25 @@
   min-height: 25px;
-  font-size: 13px;
+  font-size: 15px;
+  font-weight: bold;
```
This is a design thing, but I'm not a fan of these changes. When implementing this originally, I was very careful to make sure that the height of the `<FieldText/>` elements were the same as the height of the `<Input/>` elements. This is to ensure that when you click "Edit" on an entry you're viewing, the fields don't change position as you go from view-mode to edit-mode (unless you have fields long enough to wrap, which should be rare). By making these a different size, we lose that.

devinreams replied:
>This is to ensure that when you click "Edit" on an entry you're viewing, the fields don't change position as you go from view-mode to edit-mode

@changecourse what do you think about this consistent size so there's less jarring "jumps" in sizes between view/edit modes?

changecourse replied:
As shown here: https://mozilla.invisionapp.com/share/W3EU997YN#/262163359_1-1-2_Lockbox_Home_-_Edit_Entry

I have both the read-only text and the form field text at 15, and am trying to keep the fields from changing position to drastically, no? Toggle between the two screens for edit vs. view for context. Am I missing something?

jimporter replied:
I looked into it more and the Photon spec does say that inputs should have 15px text (ours is 13px), *but* actual Photon UI that I can find in Firefox uses 13px* (e.g. about:addons, about:preferences). Since one of the goals is to have this look like it "belongs" in Firefox, I lean towards following Firefox's lead for the time being. We should also bring this up with the Photon designers, since I'm not really clear on why there's a difference here.

* Technically 13.3167px.

jimporter commented on src/webextension/list/components/item-summary.css:
```diff
@@ -5,15 +5,38 @@
 .item-summary {
   border-bottom: 1px solid #d7d7db;
   padding: .5em;
+  font-size: 17px;
```
This looks a bit on the large size in the management UI, but my main concern here is that it looks much *more* out of place in the doorhanger. It'll also make it harder for people to see long-ish titles in the doorhanger...

jimporter commented on src/webextension/widgets/scrolling-list.css:
```diff
@@ -16,10 +16,14 @@
 .scrolling-list:focus > li[data-selected] {
-  color: #ffffff;
-  background-color: #0060df;
+  box-shadow: inset 0 0 0 2px #0a84ff, 0 0 0 3px rgba(10, 132, 255, 0.3);
```
This looks really broken on my end. The outset shadow doesn't show up properly, and the inset shadow gets cut off by the bottom border on the elements. I'm really not sure how we can fix this in a way that follows our visual spec while retaining some sort of focus state for keyboard users.

jimporter commented on src/webextension/list/components/item-list.css:
```diff
@@ -4,14 +4,23 @@
+}
+
+.item:nth-child(2n+1),
+.item:nth-child(2n) {
```
Isn't this equivalent to `.item`?

jimporter commented on src/webextension/list/manage/components/app.css:
```diff
@@ -22,35 +22,25 @@ body > main,
   display: grid;
   height: 100%;
-  grid-template-columns: 1fr 5fr;
+  grid-template-columns: 2fr 6fr;
```
Let's do `1fr 3fr`, since they're equivalent.

jimporter commented on src/webextension/list/manage/components/app.css:
```diff
@@ -22,35 +22,25 @@ body > main,
 .app-main > article {
   grid-column: 2;
-  background-color: #f9f9fa;
+  background-color: #f3f3f5;
```
Where's this color coming from? `#f9f9fa` is "Grey 10" in the Photon spec, but I don't see `#f3f3f5` anywhere. (I think I brought this up before in another file.)

devinreams replied:
>I don't see #f3f3f5

Huh.. I'm not sure where it came from. It was somewhere at some point in time..

I currently see it in 3 places and after comparing to Zeplin for this background specifically, it definitely looks like it should be replaced with #f9f9fa.

devinreams commented:
@jimporter I "put things back" the way they were in many cases based on your comments. We're finding a tough balance between "match the spec" and "intentionally diverge from it to meet our own needs". That all said, how does this look now?

I'm still looking at the two failing integration tests (`test_sign_in_with_fxa`, and `test_sign_in_with_fxa_from_home`).

devinreams commented:
@jrbenny35 Thanks again for your heroic work figuring out the craziness in this branch. We're in much better shape and with a little bit of sleep, the browser integration tests are passing somewhat consistently! ✅ 

@jimporter Can you please look again at what we have here and see if there is anything offensive enough to prevent landing? If you spot anything that can wait, please feel free to drop it here or in #392 so we don't lose it on the next pass. Thanks again for all your feedback. I learned a lot here...

@m8ttyB This is _finally_ nearly ready to rock and roll... 🎸 sorry for the delays.

devinreams commented:
Here's a quick glance at how the branch looks now.

<img width="1115" alt="screen shot 2017-12-11 at 9 26 00 am" src="https://user-images.githubusercontent.com/49511/33841669-a0178186-de55-11e7-8717-996154f9cfd8.png">

<img width="1115" alt="screen shot 2017-12-11 at 9 25 57 am" src="https://user-images.githubusercontent.com/49511/33841671-a17b834c-de55-11e7-983a-1d29b6ca8af7.png">

<img width="1115" alt="screen shot 2017-12-11 at 9 26 23 am" src="https://user-images.githubusercontent.com/49511/33841679-a39c6394-de55-11e7-8892-bde72f832403.png">




@jimporter I believe I've addressed your concerns, including the height/jump on entry save/view now (went back down through all comments).

Repeat from before but: Can you please look again at what we have here and see if there is anything offensive enough to prevent landing? If you spot anything that can wait, please feel free to drop it here or in #392 so we don't lose it on the next pass.

jimporter commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "version": "0.1.4-alpha",
   "main": "dist/bootstrap.js",
-  "description": "A Lockbox extension for Firefox",
+  "description": "the simple way to store, retrieve and manage website login info",
```
This really feels like it should have sentence case, even if the in-page string doesn't...

jimporter commented on src/webextension/widgets/button.css:
```diff
@@ -36,6 +38,39 @@
+  color: #0060df;
+  border-color: #0a84ff
+}
+
```
All this should be removed, now that we have an `<ExternalLink/>` widget.

devinreams replied:
>All this should be removed, now that we have an `<ExternalLink/>` widget.

I suppose this is all based on the above decisions re: "Sign in.." and "Copy.." being buttons with link theme vs buttons with.. button-y themes..

Point being: not all links are external which this is why we would keep this around (based on the current designs, some buttons look like links) right?

We had agreed the only places for true `<a>` links were as you see now, right (Feedback, FAQ)? Or have I misunderstood this feedback?

jimporter commented on src/webextension/widgets/link.css:
```diff
@@ -0,0 +1,36 @@
+
+.external:focus {
+  color: #0060df;
+  border-color: #0a84ff
```
Nit: add a semicolon here.

jimporter replied:
(done)

jimporter commented on src/webextension/widgets/panel.js:
```diff
@@ -11,9 +11,13 @@ import styles from "./panel.css";
   const finalClassName = `${styles.panelHeader} ${className}`.trimRight();
+  const imgSrc = browser.extension.getURL("/icons/arrowhead-left-16.svg");
+  // TODO: actually localize this string
+  const imgAlt = "gO bACk";
```
Don't forget to localize this.

devinreams replied:
>Don't forget to localize this.

@jimporter I'm not sure how. If you see my prior commit at 9db70b2 it returns "[object Object]" and I'm not sure how to get just the string out of it. Is this something you can help with?

jimporter replied:
You'll want something like the following:

widgets.ftl:
```
panel-go-back-button
  .alt = Go Back
```

panel.js:
```js
<Button theme="ghost" size="micro" onClick={onBack}>
  <Localized id="panel-go-back-button">
    <img src={imgSrc}/>
  </Localized>
</Button>
```

jimporter replied:
(done)

devinreams commented:
>Some of these things could be addressed post-release as followups, but at minimum, I'd want a guarantee that after release, we focus 100% on cleanup and fixing tech debt before we start any new features.

@jimporter Yup! We absolutely will prioritize "clean up" and addressing technical debt. Is there anything we've done so far that's suggested we won't? I think we've all done a great job acknowledging what we need to come back to and stuck to it...

This is certainly the challenge with a timeline-driven milestone: we have time-based commitments to stick to (All Hands release, elective presentation) but we can't do everything we want to in the remaining time. That said: we're a new, flexible, project and team and should be able to navigate this... 

>A few more comments below...

@jimporter Are you interested and able to help make any of these changes directly yourself?

I feel the back-and-forth review between you and I is reaching my "ability" limit (e.g.: you're much better at this and know what is best and can best address it. versus me trying to understand and make the change... with varying degrees of success).

I'll respond to each issue with what I think I need help with...

The question still remains: is anything here so offensive you would not feel comfortable merging this branch by tomorrow, early Wednesday (with a morning hack session) at the latest?

devinreams commented on src/webextension/firstrun/components/using.js:
```diff
@@ -49,7 +49,7 @@ export default function StartUsing() {
         </Localized>
         <Localized id="firstrun-using-returning-action">
-          <Button theme="ghost"
+          <Button theme="link" className="external"
```
>I think we discussed (somewhere) that this should be a button, not a link.

@jimporter You mean you believe the look and feel of "Sign in to your Firefox Account..." should be a button not a link?

<img width="392" alt="screen shot 2017-12-11 at 3 26 59 pm" src="https://user-images.githubusercontent.com/49511/33854571-c91af6a8-de87-11e7-9095-1d4223a87aa0.png">

jimporter replied:
Yeah. Somewhere else on this PR, we discussed this (I believe, you, me, and Matt M), and I argued that if we want the ARIA role to be "button", it should visually look like a button instead of a link. I see CSS and ARIA as serving similar purposes (i.e. indicating to users the "role" of an element), so it would be strange to have CSS and ARIA not match stylistically.

devinreams commented on src/webextension/list/components/item-fields.js:
```diff
@@ -21,7 +21,7 @@ const PASSWORD_DOT = "\u25cf";
   return (
     <CopyToClipboard text={text} onCopy={() => onCopy(field)}>
-      <Button {...props}/>
+      <Button theme="link" {...props}/>
```
> this should have button theming, not link theming.

@jimporter You mean you think the "Copy" text should [be a button](https://app.zeplin.io/project/59fb4868c5e7608593f76965/screen/5a1f4300904767431abe5640)?

<img width="269" alt="screen shot 2017-12-11 at 3 32 06 pm" src="https://user-images.githubusercontent.com/49511/33854779-7f262ac6-de88-11e7-84d8-12dc70de7066.png">

jimporter replied:
Right. We could give this some one-off styles, provided they don't too "link-y". A "ghost" button would probably be a good first pass.

devinreams commented on src/webextension/widgets/input.css:
```diff
@@ -78,17 +79,24 @@
   margin-top: 7px;
   min-height: 25px;
   font-size: 13px;
+  font-weight: bold;
```
>Is it ok that this is bold when the input text isn't bold? 

It is apparently intentional. Is that right @changecourse?

<img width="289" alt="screen shot 2017-12-11 at 3 40 12 pm" src="https://user-images.githubusercontent.com/49511/33855110-999ffd18-de89-11e7-9aba-a5362cf377fb.png">

devinreams commented on src/webextension/widgets/scrolling-list.css:
```diff
@@ -16,10 +16,15 @@
 .scrolling-list > li[data-selected] {
   color: #0c0c0d;
-  background-color: #b1b1b3;
+  background-color: #f9f9fa;
```
>This is the same color as even-numbered rows in the `<ItemList/>`, which is pretty confusing.

@jimporter For purposes of landing this, @changecourse suggested in slack a pure white background and selected item-summary. Shall we try that? Or something else short-term?

jimporter replied:
I don't know if that would help much, but if it looks good to you, then we can roll with it for the time being. I'm not sure I'd block the merge if this were the last issue standing though; we'll probably have to make significant style changes here in the near future anyway.

devinreams commented on src/webextension/widgets/toolbar.css:
```diff
@@ -5,6 +5,7 @@
 .toolbar {
   margin: 0;
   padding: 1ch;
+  grid-column-start: span 2;
```
>This doesn't belong here; maybe in...

I'm having a hard time with the munging and this inexact feedback.

Can you please help me learn by proposing the exact change you'd expect to see after confirming it works on-page? Or if that's a lot to test right now, we make a note in #392 to re-review?

jimporter replied:
(done)

devinreams commented on src/webextension/widgets/scrolling-list.css:
```diff
@@ -16,10 +16,15 @@
   color: #0c0c0d;
-  background-color: #b1b1b3;
+  background-color: #f9f9fa;
+  border-right: none;
```
>This should probably go in item-summary.js.

I had a hard time getting all of this sidebar / scroll list properly targeted. Are you saying the border removal itself should be removed from here specifically because its reused in the doorhanger?

jimporter replied:
Generally, yeah. The generic `<ScrollingList/>` should have a minimum of CSS (even the CSS for selected rows might be too much, but I wanted it to work ok out of the box).

devinreams commented on src/webextension/widgets/toolbar.css:
```diff
@@ -16,3 +17,8 @@
 }
+
+a,
+span {
```
>Don't use bare tag names for CSS selectors; that breaks the scoped CSS setup we're using... I'm not sure this is the right place to put this.

I don't understand the benefit of the scoping then. If we can't be liberal targeting the 1 link in the toolbar, what shall we do instead? Can you help propose the change as I'm failing to grok the abstraction..

jimporter replied:
I'll work on a patch; one of the reasons all our widgets let us pass `className` to change styles is so we can do one-off styles for stuff to make it match up. In this case, I'd apply the style in the implementation for `<SendFeedback/>`, for example.

jimporter replied:
(done)

jimporter commented:
> Is there anything we've done so far that's suggested we won't?

I just wanted to avoid any potential conflict as to what our next goals after this were. Given the choice, I'd have reduced the scope of this release so that we could keep things clean instead of landing stuff that's not-quite there and signing ourselves up for janitorial duty post-release. (A good example here is how our test coverage has been slipping, especially with my telemetry rework. I wanted to get coverage of the telemetry middleware, since it could easily break *everything* if it's wrong. Unfortunately, there wasn't time, and the telemetry stuff needed to land so that we actually get correct metrics.)

In any case, I'll start working on some patches on top of this. Since we don't exactly have real users yet, I suppose this could land anytime, but I do want to minimize the amount of post-landing cleanup we have to do, since that can be pretty tiring and we run the risk of missing stuff.

jimporter commented:
Ok, I've addressed everything except:

* [ ] Reordering the item fields (not sure if we need to do anything about this, though I do have a feeling we'll end up reverting this when we have multi-origin support)
* [ ] Bold-face or not on the non-editable item fields (just something for @changecourse to think about; see above)
* [ ] Miscellaneous small issues with the item list

jimporter commented:
@devinreams GitHub won't let me ask you for review (presumably because you're an assignee; good thing we haven't actually assigned the *entire* team to this PR yet!), so I'll just flag you manually via this fancy message! :)

devinreams commented:
Thanks for the late-night commits. This is looking great!

I'm kinda embarrassed at how easy some of the changes were (hindsight is 20/20) but the hard part is not the typing, its knowing how and where it all goes. I appreciate it. 😊 

---

>I just wanted to avoid any potential conflict as to what our next goals after this were.

Good point. And I think our best way to capture and prioritize these amongst the team's various plans is to have issues created for these added to the backlog. I agree something like "Add test coverage for Telemetry middleware" is a perfect candidate.

Then when planning time comes, we can see the various "debt" we need to pay back and it's an easy negotiation to pop those kinds of things to the top of the stack.

>Reordering the item fields (not sure if we need to do anything about this, though I do have a feeling we'll end up reverting this when we have multi-origin support)

I don't think we'll want to change this now given the Alpha user feedback. Despite there being some remaining concerns, let's try this and see if it improves the user experience as we hope it does.

Either way, yes, when we support multiple origins we'll need to re-imagine this entirely. 🙃 

I'll tack these three items onto the list in #392 anyway so we can keep them there for discussion/final decisions, either way.

---

What do you think @jimporter? Good to merge and rebase #387 and #399 against so we can release this thang? 🤠

devinreams commented on src/webextension/widgets/panel.js:
```diff
@@ -12,12 +13,15 @@ import styles from "./panel.css";
-        <img src={imgSrc} alt={imgAlt}/></Button> : null}
+      {onBack ? (
+        <Button theme="ghost" size="micro" onClick={onBack}>
+          <Localized id="panel-back-button">
```
I swear I tried this (amongst a few other things). I really get confused on how this Localized component works sometimes... thanks for this.

jimporter commented:
@devinreams I'd like to have a quick chat with Ryan first about what we can do with the item list, since I still need to do a few things there (how much I do depends on the discussion with Ryan, but the minimum amount is still greater than 0). Then we can click the big green button.

jrbenny35 commented:
Someone please add this on your next push
```
diff --git a/.travis.yml b/.travis.yml
index 7d1d1df..197b8e9 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -15,7 +15,7 @@ jobs:
       apt:
         packages:
           - jwm
-      env: TOXENV=integration-tests MOZ_HEADLESS=1 GECKODRIVER=0.19.1
+      env: TOXENV=integration-tests GECKODRIVER=0.19.1
       before_install:
         - wget -O /tmp/geckodriver.tar.gz https://github.com/mozilla/geckodriver/releases/download/v$GECKODRIVER/geckodriver-v$GECKODRIVER-linux64.tar.gz
         - mkdir $HOME/geckodriver && tar xvf /tmp/geckodriver.tar.gz -C $HOME/geckodriver
@@ -26,8 +26,6 @@ jobs:
         - nvm install node
         - pip install tox
       before_script:
-        - jwm &
-        - sleep 10
         - npm install
         - ENABLE_DOORHANGER=1 npm run package
       script: tox
```

devinreams commented:
@jimporter ItemFilter (search box) styles and blue-er focus state are looking much cleaner now.  Thank you.

@jrbenny35 and thanks again for the help with the improved integration tests.

I got over-the-shoulder approval from nearly everyone involved in this branch. We're all very ready for this to merge! I think there's plenty of good retro material in here. 😉  :ship:
